strawberryfields/tdm/tdmprogram.py

```diff
@@ -83,63 +84,136 @@ def validate_measurements(circuit, N):
     return spatial_modes
 
 
-def input_check(args, copies):
+def input_check(args):
     """Checks the input arguments have consistent dimensions.
 
     Args:
         args (Sequence[Sequence]): sequence of sequences specifying the value of the parameters
-        copies (int): number of times the circuit should be run
-
     """
     # check if all lists are of equal length
     param_lengths = [len(param) for param in args]
     if len(set(param_lengths)) != 1:
         raise ValueError("Gate-parameter lists must be of equal length.")
 
-    # check copies
-    if not isinstance(copies, int) or copies < 1:
-        raise TypeError("Number of copies must be a positive integer.")
-
 
-def _get_mode_order(num_of_values, N):
+def _get_mode_order(num_of_values, N, timebins):
     """Get the order by which the modes were measured"""
     all_modes = []
     for i in range(len(N)):
-        ra = list(range(sum(N[:i]), sum(N[: i + 1])))
-        all_modes.append(ra * ceil(max(N) / len(ra)))
+        modes = list(range(sum(N[:i]), sum(N[: i + 1])))
+        extended_modes = modes * ceil(1 + timebins // len(modes))
+        all_modes.append(extended_modes[:timebins])
 
     mode_order = [i for j in zip(*all_modes) for i in j]
-    mode_order *= ceil(num_of_values / len(mode_order))
+    mode_order *= ceil(1 + num_of_values / len(mode_order))
 
     return mode_order[:num_of_values]
 
 
-def reshape_samples(all_samples, modes, N):
+def reshape_samples(all_samples, modes, N, timebins):
     """Reshapes the samples dict so that they have the expected correct shape.
 
     Corrects the :attr:`~.Results.all_samples` dictionary so that the measured modes are the ones
     defined to be measured in the circuit, instead of being spread over a larger
     number of modes due to the mode-shifting occurring in :class:`~.TDMProgram`.
 
+    The function iterates through the unrolled circuit and populates, and returns, a new samples
+    dictionary with the shape ``{spatial mode: (shots, timebins)}``. E.g., this unrolled circuit:
+
+    .. code-block:: pycon
+
+        ...
+        MeasureHomodyne(0) | (q[0])  # shot 0, timebin 0, spatial mode 0  (sample 0)
+        MeasureHomodyne(0) | (q[2])  # shot 0, timebin 0, spatial mode 2  (sample 1)
+        ...
+        MeasureHomodyne(0) | (q[0])  # shot 0, timebin 1, spatial mode 0  (sample 2)
+        MeasureHomodyne(0) | (q[1])  # shot 0, timebin 1, spatial mode 2  (sample 3)
```

Comment:
One question here (think I'm personally just missing something): should the two target modes here be swapped for the measurements? :thinking: 

I'd think that we're shifting with one mode in timebin 1:
1. `MeasureHomodyne(0) | (q[0])` -> `MeasureHomodyne(0) | (q[1])` (spatial mode 0) 
2. `MeasureHomodyne(0) | (q[2])` -> `MeasureHomodyne(0) | (q[0])` (spatial mode 2)

And would have expected:
```
        MeasureHomodyne(0) | (q[1])  # shot 0, timebin 1, spatial mode 0  (sample 2)
        MeasureHomodyne(0) | (q[0])  # shot 0, timebin 1, spatial mode 2  (sample 3)
```

Reply:
So, in this case this would correspond to `N = [1, 2]`, i.e. one timebin in the first band and 2 timebins in the second. Since the first measurement is in the first band (which only has timebin 0), it stays there:

`MeasureHomodyne(0) | (q[0]) -> MeasureHomodyne(0) | (q[0])`

The second measurement would start on band 2 (which has timebins 1 and 2), and would loop around to timebin 1:

`MeasureHomodyne(0) | (q[2]) -> MeasureHomodyne(0) | (q[1])`

You can check it by running the following:

```pycon
>>> with prog.context([1,2],[4,5]) as (p, q): 
...     BSgate(p[0]) | (q[0], q[1]) 
...     Rgate(p[1]) | q[1] 
...     MeasureHomodyne(p[0]) | q[0] 
...     MeasureHomodyne(p[1]) | q[2] 

>>> eng = sf.Engine("gaussian") 
>>> shots = 2 
>>> result = eng.run(prog, shots=shots)                                    

>>> for c in prog.unrolled_circuit: print(c)
                            
BSgate(1, 0) | (q[0], q[1])
Rgate(4) | (q[1])
MeasureHomodyne(1) | (q[0])
MeasureHomodyne(4) | (q[2])
BSgate(2, 0) | (q[0], q[2])
Rgate(5) | (q[2])
MeasureHomodyne(2) | (q[0])
MeasureHomodyne(5) | (q[1])
```

But this is just one example of the program, and this docstring example doesn't actually tell which circuit we're running (since it's only concerned about the reshaping). :slightly_smiling_face: 

The `...`s would be the other gates in the circuit, that also aren't important for the reshaping.

Reply:
Oh got ya, thank you!

When `shots=2`, got the following error though:
```
~/xanadu/strawberryfields/strawberryfields/engine.py in _run_program(self, prog, **kwargs)
    371                         cmd.op, self.backend, kwargs
    372                     )
--> 373                 ) from None
    374 
    375         samples = self._combine_and_sort_samples(samples_dict)

NotImplementedError: The operation MeasureHomodyne(1) has not been implemented in GaussianBackend for the arguments {'shots': 2}.
```

Reply:
Are you running it on the correct branch? That error shouldn't be happening (since, for a TDMProgram, it should set shots to 1 before moving on after unrolling the circuit in `eng.run`).

Reply:
You're right, was my bad! Got confused with branch names